src/util_types/mutator_set/ms_membership_proof.rs

```diff
@@ -99,24 +96,55 @@ impl MsMembershipProof {
             "Function must be called with same number of membership proofs and items. Got {} items and {} membership proofs", own_items.len(), membership_proofs.len()
         );
 
-        let new_item_index = mutator_set.aocl.count_leaves();
+        assert!(membership_proofs
+            .iter()
+            .all(|msmp| msmp
+                .target_chunks
+                .iter()
+                .all(|(chunk_index, (mmr_mp, chunk))| mmr_mp.verify(
+                    *chunk_index,
+                    Hash::hash(chunk),
+                    &mutator_set.swbf_inactive.peaks(),
+                    mutator_set.swbf_inactive.num_leafs()
+                ))));
+
+        let new_item_index = mutator_set.aocl.num_leafs();
 
         // Update AOCL MMR membership proofs
-        let indices_for_updated_mps = MmrMembershipProof::batch_update_from_append(
+        let leaf_indices = membership_proofs
+            .iter()
+            .map(|msmp| msmp.aocl_leaf_index)
+            .collect_vec();
+        let indices_for_mps_updated_from_append = MmrMembershipProof::batch_update_from_append(
             &mut membership_proofs
                 .iter_mut()
-                .map(|x| &mut x.auth_path_aocl)
+                .map(|msmp| &mut msmp.auth_path_aocl)
                 .collect::<Vec<_>>(),
+            &leaf_indices,
             new_item_index,
             addition_record.canonical_commitment,
-            &mutator_set.aocl.get_peaks(),
+            &mutator_set.aocl.peaks(),
         );
 
         // if window does not slide, we are done
         if !MutatorSetAccumulator::window_slides(new_item_index) {
-            return Ok(indices_for_updated_mps);
+            return Ok(indices_for_mps_updated_from_append);
         }
 
+        assert!(membership_proofs
```

Comment:
Again?

Reply:
Thanks for spotting. I'm removing this one.